### PR TITLE
TrackSelection: fix wrong calculation of dca

### PR DESCRIPTION
**Pull request by mario-krueger**

### CODE REVIEW

**Comment by mario-krueger** [comment]
The calculation of DCA was done wrong before.
The version I upload now produces exactly the same results as in Run2, however since we calculate the dca for all (also the broken) tracks this now results in lots of warnings for the ones where the propagations fail emitted from line 362 here:
https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/Reconstruction/src/Track.cxx
Therefore, I am not sure if we can merge it like this. Eventually we need a way to skip calculation of this quantity (and others) for the cases where this is known to fail. Comments and Ideas would be welcome.

**Comment by jgrosseo** [comment]
@shahor02 what do you think?

File: Analysis/Tasks/trackselection.cxx [hunk 1]
```diff
@@ -74,25 +77,17 @@ struct TrackExtensionTask {
...
-      dcaXY = track.charge() * sqrt(pow((globalX - collision.posX()), 2) +
-                                    pow((globalY - collision.posY()), 2));
-      dcaZ = track.charge() * sqrt(pow(track.z() - collision.posZ(), 2));
-
-      extendedTrackQuantities(dcaXY, dcaZ);
+      // FIXME: can we simplify this knowing that track is already at dca without copying code from TrackPar?
+      float magField = 5.0; // in kG (FIXME: get this from CCDB)
+      std::array<float, 2> dca{};
+      auto trackPar = getTrackParCov(track);
+      trackPar.propagateParamToDCA({collision.posX(), collision.posY(), collision.posZ()}, magField, &dca);
```
Comment (shahor02) [inline Analysis/Tasks/trackselection.cxx]: Yes, either init ``dca`` with very large values or check the returned value of ``propagateParamToDCA`` and then set dca to large value.

File: Analysis/Tasks/trackselection.cxx [hunk 2]
```diff
@@ -74,25 +77,17 @@ struct TrackExtensionTask {
...
-
-      dcaXY = track.charge() * sqrt(pow((globalX - collision.posX()), 2) +
-                                    pow((globalY - collision.posY()), 2));
-      dcaZ = track.charge() * sqrt(pow(track.z() - collision.posZ(), 2));
-
-      extendedTrackQuantities(dcaXY, dcaZ);
+      // FIXME: can we simplify this knowing that track is already at dca without copying code from TrackPar?
+      float magField = 5.0; // in kG (FIXME: get this from CCDB)
+      std::array<float, 2> dca{};
+      auto trackPar = getTrackParCov(track);
```
Comment (shahor02) [inline Analysis/Tasks/trackselection.cxx]: For operations on Track params only you don't need to create full fledged TrackParCov, I think a method ``getTrackPar`` should be added to trackUtilities.

**Comment by shahor02** [comment]
@mkruegerGitHub @jgrosseo if the ``track::propagateParamToDCA`` returns false (the warnings will be eventually changed to debug) it does not mean the track is "broken" (there are no such tracks), it just means that either the DCA is > ``maxD`` argument or that to reach the DCA the track has to go outside of its validity range (of local phi<+-pi/2). Usually this means that the track is a low-pt looper. Technically, it can be propagated to PCA by doing an additional rotation, but do we need this?

**Comment by mario-krueger** [comment]
@shahor02 Thanks for the explanation! I think we only need the dca in this case.
If we could change the printout from warning to debug this would probably already help to have a working version that we could commit.
Is a chi2_(its,tpc) =0 an indication that the track is a low-pt looper?
If such checks can be performed to skip the calculation of dca overall this would be probably making things a bit faster.
Having a more 'ligthweight' option to obtain the dca as you mention in your comment would indeed also be great :)

**Comment by mario-krueger** [comment]
I will add getTrackPar to the utilities

**Comment by shahor02** [comment]
@mkruegerGitHub if it really hurts to have these warnings I can change them, but at the development phase I would prefer to have some verbosity, let me know.

For the loopers: no, chi2 has nothing to do with them, do you really see chi2=0 in converted data? There should be no such tracks. In general: for tracks seen in the ITS the failure rate in propagateParamToDCA with primary vertex should be extremely low. For TPC-only tracks a field dependent pT cut-off (~100 MeV at 5kG) should eliminate most of tracks prone to fail.
But the TPC-only tracks should be used with great care in O2 analyses (I am even not sure we will export all of them to AOD), I would not put them into the same loop as tracks with ITS.

**Comment by mario-krueger** [comment]
I see plenty of chi2=0 tracks. But now that you mention it, most likely the problem is that I dont filter the track type before trying to calculate the dca so these tracks might well be ﻿﻿ITSStandalone, MFTStandalone or Run2Tracklet.
I will check. Maybe this already solves the problem. Thanks!

**Comment by mario-krueger** [comment]
Unfortunately this did not help. There are still many tracks with chi2=0.

**Comment by shahor02** [comment]
This should mean that there is a problem inchi2 definition of converted track, all barrel tracks must have non0 chi2

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by mario-krueger** [comment]
Okay then we schould probably revisit the converter.
Also apart from the chi2=0 cases, there are still 191 (in 85 PbPb events) remaining tracks that can not be propagated.
For these tracks it seems that they all have very large x or z values (1.5m-2m).

**Comment by mario-krueger** [comment]
This is a bit of a hack, but temporarily fixes the problem with the converted data we have now.
I am losing only one 'good' track (according to run2 track selection) due to the track.tpcChi2NCl() != 0.f requirement.

**Comment by shahor02** [comment]
> Okay then we schould probably revisit the converter.
> Also apart from the chi2=0 cases, there are still 191 (in 85 PbPb events) remaining tracks that can not be propagated.
> For these tracks it seems that they all have very large x or z values (1.5m-2m).

these should be lot pT TPC tracks which failed propagation to vertex already in the aliroot.

**Comment by jgrosseo** [comment]
> But the TPC-only tracks should be used with great care in O2 analyses (I am even not sure we will export all of them to AOD), I would not put them into the same loop as tracks with ITS.

For Run 2 conversion we convert all tracks. For Run 3, I think it is fine if there are no TPC only tracks, but this we have to discuss also with CF. They are worried about holes in the acceptance, but I think this problem will also reduce drastically with the new ITS.

**Pull request merged**